File: apps/mobile/lib/shared/flat_picker.dart
import '../api/models.dart';

String wingKey(FlatDto flat) => (flat.wingName ?? '').trim();

const unnamedWingToken = '__unnamed__';

String toWingSelectValue(String wing) =>
    wing.isEmpty ? unnamedWingToken : wing;

String fromWingSelectValue(String value) =>
    value == unnamedWingToken ? '' : value;

String wingLabel(String wing) => wing.isEmpty ? 'No wing' : wing;

List<String> uniqueWingNames(List<FlatDto> flats) {
  final seen = <String>{};
  final names = <String>[];
  for (final flat in flats) {
    final key = wingKey(flat);
    if (seen.add(key)) names.add(key);
  }
  names.sort();
  return names;
}

List<FlatDto> flatsInWing(List<FlatDto> flats, String wing) {
  final rows = flats.where((flat) => wingKey(flat) == wing).toList()
    ..sort((a, b) => a.number.compareTo(b.number));
  return rows;
}

String? firstFlatIdInWing(List<FlatDto> flats, String wing) {
  final rows = flatsInWing(flats, wing);
  if (rows.isEmpty) return null;
  return rows.first.id;
}

String wingForFlatId(List<FlatDto> flats, String? flatId) {
  for (final flat in flats) {
    if (flat.id == flatId) return wingKey(flat);
  }
  final names = uniqueWingNames(flats);
  return names.isEmpty ? '' : names.first;
}

File: packages/ui/src/flat-picker.ts
import type { FlatDto } from "@society-hub/types";

export type WingFlatRow = Pick<FlatDto, "id" | "number" | "wingName">;

export function wingKey(flat: Pick<WingFlatRow, "wingName">): string {
  return (flat.wingName ?? "").trim();
}

export function wingLabel(wing: string): string {
  return wing || "No wing";
}

export function uniqueWingNames(flats: WingFlatRow[]): string[] {
  const seen = new Set<string>();
  const names: string[] = [];
  for (const flat of flats) {
    const key = wingKey(flat);
    if (seen.has(key)) continue;
    seen.add(key);
    names.push(key);
  }
  return names.sort((a, b) => a.localeCompare(b, undefined, { numeric: true }));
}

export function flatsInWing(flats: WingFlatRow[], wing: string): WingFlatRow[] {
  return flats
    .filter((flat) => wingKey(flat) === wing)
    .sort((a, b) =>
      a.number.localeCompare(b.number, undefined, { numeric: true }),
    );
}

export function firstFlatIdInWing(flats: WingFlatRow[], wing: string): string {
  return flatsInWing(flats, wing)[0]?.id ?? "";
}

export function wingForFlatId(flats: WingFlatRow[], flatId: string): string {
  const selected = flats.find((flat) => flat.id === flatId);
  return selected ? wingKey(selected) : (uniqueWingNames(flats)[0] ?? "");
}
